discord/client.py

```diff
@@ -1128,10 +1125,46 @@ async def on_ready():
         if not asyncio.iscoroutinefunction(coro):
             raise TypeError("event registered must be a coroutine function")
 
-        setattr(self, coro.__name__, coro)
+        self.events[coro.__name__] = coro
         _log.debug("%s has successfully been registered as an event", coro.__name__)
         return coro
 
+    def event_for(self, event_name: Union[str, EventType]):
+        """Similar to :meth:`@Client.event` except that it uses a provided name instead
+        of the name for the coroutine.
+
+        Example
+        ---------
+
+        .. code-block:: python3
+
+            @client.event_for(EventType.message)
+            async def message_here(message):
+                print(message.content)
+
+        Parameters
+        ------------
+        event_name: :type:`Union[str, EventType]`
+            The name of the event to register this coroutine under.
+
+        Raises
+        --------
+        TypeError
+            The coroutine passed is not actually a coroutine.
+        """
+        def wrapped(coro: Coro):
+            if not asyncio.iscoroutinefunction(coro):
+                raise TypeError("event registered must be a coroutine function")
+
+            actual_event = str(event_name)
+            if not actual_event.startswith("on_"):
+                actual_event = "on_" + actual_event
+
+            self.events[actual_event] = coro
```

Comment:
Maybe store the event without the `on_`, or store it under the respective enum

Reply:
The thing is, I still want for this system to work with custom events that the developer might have, so storing it under the respective enum won't work. Hence why the `event_for` and `dispatch` functions still accept `str` types for the event names.